zabbix_auto_config/processing.py

```diff
@@ -301,7 +309,6 @@ def merge_sources(self):
             for host_modifier in self.host_modifiers:
                 try:
                     modified_host = host_modifier["module"].modify(host.copy(deep=True))
-                    assert isinstance(modified_host, models.Host)
```

Comment:
I thought the protocol `isinstance` check also checked method signatures, but apparently it doesn't, so you are right that it's not technically redundant. 

I can put that assertion back, but I think ideally returning a non-host object from a modifier should be a fatal error instead of just an error message?  The error will likely recur for every single host, and as the module breaking the interface established by ZAC, we should refuse to run as long as the problematic modifier is present IMO. Thoughts?

Reply:
(For the record: I'm not at all familiar with `Protocol`, but by creating a simple test it didn't seem to check the signature, yes).

I.e. raise a `ZACException` and crash? I think that's fair. If you have a "rogue modifier" you might have lost control and it's probably safest to fully fail/crash. I was probably wondering about how to do this when writing it (the `# TODO: Do more?` comment).

Reply:
If we really want to perform some more in-depth type checking of the modules, we could use a function that looks something like this:

```py
def check_modifier_type(modifier: object, name: str) -> None:
    """Ensures the host modifier module object is valid."""
    if not isinstance(modifier, HostModifierModule):
        raise TypeError(
            f"Host modifer module {name!r} does not implement the HostModifierModule protocol. Check the docs for more info."
        )

    sig = modifier.modify.__annotations__
    if not sig:
        logging.warning(
            "Host modiifer module '%s' has no type annotations. Unable to ensure type safety.",
            name,
        )
        return

    # Check type annotations from here on
    rtn = sig.pop("return", None)
    if rtn and rtn != Host:  # Check return annotation separately
        raise TypeError(f"{name}.modify must return a Host object")
    if len(sig) != 1:
        raise TypeError(
            f"{name}.modify must take exactly one argument, function takes {len(sig)} arguments."
        )
    arg = next(iter(sig.values()))
    if arg != Host:
        raise TypeError(f"The first argument of {name}.modify must be a Host object")
```

This comes with the caveat of being unable to properly check untyped code, but should give nice error messages for typed code.

I'll re-add the assertion, but I think assertions can be a bit dangerous with regards to `-O` / `PYTHONOPTIMIZE`, but that's an [ongoing discussion](https://discuss.python.org/t/stop-ignoring-asserts-when-running-in-optimized-mode/13132), and we could just state up-front that the application should not be run in optimized mode.

Reply:
> I.e. raise a ZACException and crash?

We will discuss this further in a meeting tomorrow, but for now I'm actually walking back on my suggestion. We are in the process of looking at integrating sources/modifiers from a 3rd party, and we don't want to give them the ability to lock up ZAC due to errors in their code. Instead, ZAC should just deactivate that source/modifier permanently until it is restarted, and then log the error which we can detect and trigger an alarm for.

Reply:
It's hard to foretell what might happen if a source and/or modifier becomes unavailable or creates malformed results. Crashing is a much better situation than destruction of data/state, but there is always the configured failsafe, so I don't know...

Reply:
We have decided that failing modifiers should log an error instead of crashing, and then be disabled upon X failures in Y seconds. A future PR will implement this for both host modifiers and source collectors. This PR is ready to be merged.

Reply:
I'd really suggest against doing that, since you're losing control, but oh well.